### PR TITLE
feat(rust/rolldown): support `inject` imports

### PR DESCRIPTION
**Pull request by Boshen**
This PR adds the missing pieces for inject plugin.

### CODE REVIEW

File: crates/rolldown/src/utils/pre_process_ecma_ast.rs [hunk 1]
```diff
@@ -23,6 +25,19 @@ pub fn pre_process_ecma_ast(
...
 ) -> anyhow::Result<(EcmaAst, SymbolTable, ScopeTree)> {
+  // Build initial semantic data and check for semantic errors.
+  let semantic_ret = ast.program.with_mut(|WithMutFields { program, source, .. }| {
+    SemanticBuilder::new(source, source_type).build(program)
+  });
+
+  // TODO:
+  // if !semantic_ret.errors.is_empty() {
+  // return Err(anyhow::anyhow!("Semantic Error: {:#?}", semantic_ret.errors));
+  // }
```
Comment (Boshen) [inline crates/rolldown/src/utils/pre_process_ecma_ast.rs:38]: This broke a few test cases so I left this as a TODO.

File: crates/rolldown/src/utils/pre_process_ecma_ast.rs [hunk 2]
```diff
@@ -43,21 +58,37 @@ pub fn pre_process_ecma_ast(
...
+    // Use built-in inject plugin.
+    // <https://www.npmjs.com/package/@rollup/plugin-inject>
+    // TODO: move config to Rolldown options
+    // NOTE: `InjectGlobalVariablesConfig` is `Arc`ed.
+    // let config = InjectGlobalVariablesConfig::new(vec![InjectImport::named_specifier(
+    // "es6-promise",
+    // [> "default" <] None,
+    // "Promise",
+    // )]);
+    // InjectGlobalVariables::new(allocator, config).build(&mut symbols, &mut scopes, program);
```
Comment (Boshen) [inline crates/rolldown/src/utils/pre_process_ecma_ast.rs]: Please uncomment this block to get things going.

**Comment by hyf0** [comment]
Any idea about why performance get down so much even we don't use inject? 

---edited

Besides adding skipping logic, even we do a full-ast traverse with empty options, it shouldn't slow that much. [This PR](https://github.com/rolldown/rolldown/pull/1927) adds a almost full-ast traverse pass, but have little affection.

File: crates/rolldown/tests/rolldown/function/inject/main.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import assert from 'node:assert'
+
+assert.strictEqual(Promise, 'promise-shim')
+assert.strictEqual(P, 'promise-shim')
+assert.strictEqual($, 'jquery')
+assert.strictEqual(fs.default, 'node-fs')
+// FIXME: oxc injects invalid statements `import { default as 'Object.assign' from 'object-assign-shim'` }`, so it fails.
```
Comment (hyf0) [inline crates/rolldown/tests/rolldown/function/inject/main.js:7]: @Boshen cc

There is a bug here.

I take a rough look

oxc transform to the following code

```js
import { default as 'Object.assign'} from 'object-assign-shim'`
```

This is invalid js code.

while it should be 

```js
import { default as Object_assign} from 'object-assign-shim'`
```
or
```js
import Object_assign from 'object-assign-shim'`
```

**Review by hyf0 — APPROVED** [review]

**Comment by Boshen** [comment]
> Any idea about why performance get down so much even we don't use inject?
> 
> ---edited
> 
> Besides adding skipping logic, even we do a full-ast traverse with empty options, it shouldn't slow that much. [This PR](https://github.com/rolldown/rolldown/pull/1927) adds a almost full-ast traverse pass, but have little affection.

This is because semantic is now run for each file in this file, where previously its only run for ts or jsx files.

File: crates/rolldown/tests/rolldown/function/inject/main.js (hunk 3 above)
Comment (Boshen) [reply crates/rolldown/tests/rolldown/function/inject/main.js:7]: https://github.com/oxc-project/oxc/issues/4818

**Pull request merged**